### PR TITLE
[PLAT-761] markdown-it-mfr

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

We want to spin-off the wiki markdown-it-video `@[osf](....)` tag into it's own plugin with more flexible regex for the staging server etc.

Also there's a usability issue where the wiki markdown become particularly aggravating because pressing backspace sends you back a page in Firefox. 

## Changes

- pass internal domain and mfr domain to js contexVars
- disable backspace sending you  back a page
- adds markdown-it-mfr and downgrades markdown-it-video to a version with no `@[osf](....)` tag. 
- adds to yarn.lock

## QA Notes

This should allow `@[osf](....)` tags on staging to only staging and `@[osf](....)` tags prod to only access prod.

Also you should be able to click outside of the textbox and press backspace and not get sent back a page.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/PLAT-761

### CODE REVIEW

File: website/static/js/markdown.js [hunk 1]
```diff
@@ -30,12 +31,18 @@ var bootstrapTable = function(md) {
...
 
 // Full markdown renderer for views / wiki pages / pauses between typing
 var markdown = new MarkdownIt('commonmark', {
     highlight: highlighter,
     linkify: true
-})
+    }).use(require('markdown-it-mfr'), {
+        type: 'osf',
+        pattern: /^http(?:s?):\/\/(?:www\.)?[a-zA-Z0-9 .:]{1,}\/render\?url=http(?:s?):\/\/[a-zA-Z0-9 .:]{1,}\/([a-zA-Z0-9]{5})\/\?action=download|(^[a-zA-Z0-9]{5}$)/,
+        formatUrl: function(assetID) { return mfrURL + 'render?url='+ osfURL + assetID + '/?action=download%26mode=render'; },
```
Comment (pattisdr) [inline website/static/js/markdown.js]: Why is &mode=render needed here?

File: website/static/js/markdown.js [hunk 2]
```diff
@@ -30,12 +31,18 @@ var bootstrapTable = function(md) {
...
+var osfURL = window.contextVars.osfURL;
 
 // Full markdown renderer for views / wiki pages / pauses between typing
 var markdown = new MarkdownIt('commonmark', {
     highlight: highlighter,
     linkify: true
-})
+    }).use(require('markdown-it-mfr'), {
+        type: 'osf',
+        pattern: /^http(?:s?):\/\/(?:www\.)?[a-zA-Z0-9 .:]{1,}\/render\?url=http(?:s?):\/\/[a-zA-Z0-9 .:]{1,}\/([a-zA-Z0-9]{5})\/\?action=download|(^[a-zA-Z0-9]{5}$)/,
```
Comment (pattisdr) [inline website/static/js/markdown.js]: When trying to embed a file that doesn't exist, whose guid has five-characters,` @[osf](12345)`, you see a NOT FOUND message like you do in prod.  However, if your incorrect guid has more or less than 5 characters @[osf](1234), it doesn't attempt to make a request to mfr, it just renders it as `@osf`, where the osf is a link, and when clicked, leads to `/<node_id>/wiki/<wiki%20name>/<4 char incorrect guid>/`. This deviates from prod behavior where a 4 or 6 character guid would be requested from MFR, and a Not Found would be displayed.

Comment (Johnetordoff) [reply website/static/js/markdown.js]: I actually think this is a little better because if the guid is wrong it fails imeddiately instead of loading MFR just to say you got the guid wrong.

It's difficult to fix the @osf issue because the valid markdown for linking is `[label](link)` so when the guid isn't valid it veers into making it a link with a @ before the label. Unsure what to do about that, it's sort of a product question.

File: website/static/js/markdown.js (hunk 1 above)
Comment (Johnetordoff) [reply website/static/js/markdown.js]: Not exactly sure, but when it's removed MFR errors with the following: ` aiohttp.errors.ClientOSError: [Errno None] Cannot connect to host localhost:7777 ssl:False [Can not connect to localhost:7777 [None]]`

File: website/static/js/markdown.js (hunk 2 above)
Comment (pattisdr) [reply website/static/js/markdown.js]: I agree it's better to not make a request to MFR if we don't have to but are we 100% sure that this 5-char guid assumption always holds? I seem to remember that working on data migrations in the past, on prod, sometimes old stuff doesn't have 5 chars? Could be misremembering or this might not be true for file guids.

Comment (pattisdr) [reply website/static/js/markdown.js]: Also, can you run this by Product @Johnetordoff, and make sure that they are fine with this new behavior?

Comment (Johnetordoff) [reply website/static/js/markdown.js]: So I thought about this some more and I think your initial thoughts are correct. Having testing it using >5 character guids, it returns a "NOT FOUND" message quickly enough that it's actually more helpful than a dead link, so I'm just going to update the regex.

Comment (pattisdr) [reply website/static/js/markdown.js]: Sounds good @Johnetordoff

**Comment by sloria** [comment]
https://github.com/Johnetordoff/markdown-it-mfr/blob/5d59cf004a5a71a9367fd03b1369379ded77708b/index.js#L95 

Will this conflict with our video plugin in any way?

**Comment by sloria** [comment]
This will generate a random float, which will result in a strange ID: https://github.com/Johnetordoff/markdown-it-mfr/blob/5d59cf004a5a71a9367fd03b1369379ded77708b/index.js#L74

How about using a timestamp? Something like

```javascript
var id = '__markdown-it-mfr-' + (new Date()).getTime();
// "__markdown-it-mfr-1527603178641"
```

**Comment by Johnetordoff** [comment]
@sloria For this link https://github.com/Johnetordoff/markdown-it-mfr/blob/5d59cf004a5a71a9367fd03b1369379ded77708b/index.js#L76, you're looking at the wrong commit, you should be looking at develop where you can see more then that line has been changed.

It doesn't conflict with markdown-it-video in any obvious way.

**Comment by sloria** [comment]
> you should be looking at develop where you can see more then that line has been changed.

Ok, looked at the latest commit on develop. It still looks like https://github.com/Johnetordoff/markdown-it-mfr/blob/develop/index.js#L79 is the only MFR-specific line and that the use case is met by the more general API suggested in the ticket.

**Comment by Johnetordoff** [comment]
I think you can see the generalization here: https://github.com/Johnetordoff/markdown-it-mfr/blob/develop/index.js#L30-L35. I think I followed the ticket implementation pretty closely. 

Maybe it's confusing if you look at the regex [here](https://github.com/CenterForOpenScience/osf.io/pull/8250/files?utf8=%E2%9C%93&diff=unified#diff-a18d385960f6b4c82ca2ed6f3799ac73R43) it appears different then the ticket, but that's just because the ticket doesn't include the regex for the link, only the guid, so the pattern regex I used looks very different, but it's fundamentally the same.

File: website/static/js/markdown.js (hunk 2 above)
Comment (sloria) [reply website/static/js/markdown.js]: > Edit: Also +5 charecter guids are forward compatible, so I actually think it's not really a product thing.

@Johnetordoff Do you mean that the current code handles >5-character GUIDS? It looks like the regex still only captures 5 characters (`(^[a-zA-Z0-9]{5}$)`). Or am I misunderstanding?

**Comment by sloria** [comment]
Also, before this is merged, the repo will need to be transferred to CenterForOpenScience and published to npm. Once you've addressed the above suggestions, you can transfer the repo to me (sloria) and I can transfer it to COS.

**Comment by Johnetordoff** [comment]
Time-stamping the id here creates identical ids because the date class is evaluated when markdown-it is configured, not when the embed is created, so I added the random component back in.

**Comment by sloria** [comment]
> Time-stamping the id here creates identical ids because the date class is evaluated when markdown-it is configured, not when the embed is created, so I added the random component back in.

I'm confused about this. Shouldn't `Date()` be called every time `format` is called?

File: website/static/js/markdown.js [hunk 3]
```diff
@@ -30,12 +31,26 @@ var bootstrapTable = function(md) {
...
 
 // Full markdown renderer for views / wiki pages / pauses between typing
 var markdown = new MarkdownIt('commonmark', {
     highlight: highlighter,
     linkify: true
-})
+    }).use(require('markdown-it-atrules'), {
+        type: 'osf',
+        pattern: /^http(?:s?):\/\/(?:www\.)?[a-zA-Z0-9 .:]{1,}\/render\?url=http(?:s?):\/\/[a-zA-Z0-9 .:]{1,}\/([a-zA-Z0-9]{1,})\/\?action=download|(^[a-zA-Z0-9]{1,}$)/,
+        format(assetID) {
```
Comment (sloria) [inline website/static/js/markdown.js]: Drive-by: We don't transpile ES6, so you can't use ES6-style methods.

File: website/static/js/markdown.js [hunk 4]
```diff
@@ -30,12 +31,26 @@ var bootstrapTable = function(md) {
...
 // Full markdown renderer for views / wiki pages / pauses between typing
 var markdown = new MarkdownIt('commonmark', {
     highlight: highlighter,
     linkify: true
-})
+    }).use(require('markdown-it-atrules'), {
+        type: 'osf',
+        pattern: /^http(?:s?):\/\/(?:www\.)?[a-zA-Z0-9 .:]{1,}\/render\?url=http(?:s?):\/\/[a-zA-Z0-9 .:]{1,}\/([a-zA-Z0-9]{1,})\/\?action=download|(^[a-zA-Z0-9]{1,}$)/,
+        format: function(assetID) {
+          var id = '__markdown-it-atrules-' + (new Date()).getTime() + Math.random() ;
```
Comment (sloria) [inline website/static/js/markdown.js]: Does it make a difference if you use `Date.now()` instead of `(new Date()).getTime())`?

Comment (sloria) [reply website/static/js/markdown.js]: I'd like to avoid `Math.random()` if possible. But if `Math.now()` doesn't fix the issue, just make sure to append a random integer instead of a random float. We don't want to have `.` characters in CSS IDs because they break CSS selectors (e.g. `#__markdown-it-atrules-1234.56` would select an element with ID `#__markdown-it-atrules-1234` and class `56`).

Comment (Johnetordoff) [reply website/static/js/markdown.js]: I've tried `now()` and used the moment.js library The problem seems to be that every time something is added the div is redrawn fast enough that they're the same to the millisecond.

File: website/static/js/markdown.js [hunk 5]
```diff
@@ -30,12 +31,28 @@ var bootstrapTable = function(md) {
...
 // Full markdown renderer for views / wiki pages / pauses between typing
 var markdown = new MarkdownIt('commonmark', {
     highlight: highlighter,
     linkify: true
-})
+    }).use(require('markdown-it-atrules'), {
+        type: 'osf',
+        pattern: /^http(?:s?):\/\/(?:www\.)?[a-zA-Z0-9 .:]{1,}\/render\?url=http(?:s?):\/\/[a-zA-Z0-9 .:]{1,}\/([a-zA-Z0-9]{1,})\/\?action=download|(^[a-zA-Z0-9]{1,}$)/,
+        format: function(assetID) {
+             var id = '__markdown-it-atrules-' + mfrId++;
```
Comment (Johnetordoff) [inline website/static/js/markdown.js]: @sloria  This works because every time you add a tag it redraws the all other tags, so because mfrId is always incrementing even after you save your changes the ids get redrawn and  you can never repeat the same ID for tags in the same wiki. 

I initially didn't think this would work in all cases, but it seems to work with saved tags and collaborative edits so I'm sticking with this because it seems to guarantee uniqueness.

Comment (sloria) [reply website/static/js/markdown.js]: > collaborative edits

Did you test this with multiple users? I would think that this would break with two users, because their counters would both start at 0, so it would be possible to have duplicate IDs.

Comment (sloria) [reply website/static/js/markdown.js]: >  because every time you add a tag it redraws the all other tags, 

Ah, I didn't read this closely enough. OK, cool, then yeah, I suppose this would work!

Comment (sloria) [reply website/static/js/markdown.js]: ^ and that explains why `Date()` didn't work. If everything gets rendered at the same time, then all the timestamps will the same. So `format` *was* getting called multiple times, but it was called multiple times at the same time. Whew, I'm glad there's an explanation.

**Comment by Johnetordoff** [comment]
This passed on my branch https://travis-ci.org/Johnetordoff/osf.io/builds/391434410
